File: 43-multiply-strings/multiply-strings.py
class Solution:
    def multiply(self, num1: str, num2: str) -> str:
        if num1 == "0" or num2 == "0":
            return "0"

        # Make num1 the longer one so it is easier to reason about
        if len(num1) < len(num2):
            num1, num2 = num2, num1
        
        # Max number of digits in the result
        result = [0 for _ in range(len(num1) * 2 + 1)]

        for j in range(len(num2) - 1, -1, -1):
            for i in range(len(num1) - 1, -1, -1):
                # Initialize relevant indexes and values
                digit_1, digit_2 = int(num1[i]), int(num2[j])
                digit_1_index, digit_2_index = len(num1) - 1 - i, len(num2) - 1 - j
                res_index = digit_1_index + digit_2_index

                # Calculation
                res = digit_1 * digit_2 % 10
                c = (digit_1 * digit_2 - res) // 10

                # Adding to result the res and carry
                result[res_index] += res
                result[res_index + 1] += c

        # Roll forward the overflows, i.e. Addition portion
        for i in range(len(result)):
            overflowed_val = result[i]
            result[i] = overflowed_val % 10
            cur_carry = (overflowed_val - result[i]) // 10
            if i != len(result) - 1:
                result[i + 1] += cur_carry

        # Final processing
        i = len(result) - 1
        while result[i] == 0:
            i -= 1

        result = result[:i + 1]
        result.reverse()

        return "".join([str(num) for num in result])

File: 43-multiply-strings/multiply-strings.cpp
class Solution {
public:
    string multiply(string num1, string num2) {
        if (num1 == "0" || num2 == "0") {
            return "0";
        }

        if (num1.size() < num2.size()) {
            std::swap(num1, num2);
        }

        std::vector<int> res (num1.size() + num2.size() + 1);
        std::reverse(num1.begin(), num1.end());
        std::reverse(num2.begin(), num2.end());

        for (int i{}; i < num2.size(); ++i) {
            for (int j{}; j < num1.size(); ++j) {
                int num1digit {num1[j] - '0'};
                int num2digit {num2[i] - '0'};
                int sum {num1digit * num2digit};
                int remain {sum % 10};
                int carry {sum / 10};
                res[i + j] += remain;
                res[i + j + 1] += carry;
            }
        }

        for (int i{}; i < res.size(); ++i) {
            int val {res[i]};
            res[i] = val % 10;
            if (i != res.size() - 1) {
                res[i + 1] += val / 10; 
            }
        }
        
        std::string res_string {};
        bool leadingZeros {true};
        for (int i = res.size() - 1; i >= 0; --i) {
            if (res[i] != 0) {
                leadingZeros = false;
                res_string += static_cast<char>(res[i] + '0');
            } else if (!leadingZeros) {
                res_string += static_cast<char>(res[i] + '0');
            }
        }

        return res_string;
    }
};